File: Python/BOJ/1992 쿼드트리.py
def tree(y,x,n):
    global result
    color = data[y][x]
    for i in range(y,y+n):
        for j in range(x,x+n):
            if color != data[i][j]:
                m = n//2
                result += '(' 
                tree(y, x, m) # 왼쪽 위 (2사분면)
                tree(y, x+m, m) # 오른쪽 위 (1사분면)
                tree(y+m, x, m) # 왼쪽 아래 (3사분면)
                tree(y+m, x+m, m) # 오른쪽 아래 (4사분면)
                result += ')'
                return
    if color == 1: # 영역이 1이면
        result += str(1) # 1 표시
    else : # 영역이 0 이면
        result += str(0) # 0 표시

import sys
N = int(sys.stdin.readline())
data = [list(map(int, sys.stdin.readline().strip())) for _ in range(N)]
result = ''
tree(0,0,N)
print(result)

File: Java/BOJ/1992 쿼드트리.java
import java.io.*;
import java.util.*;

public class Main {
	static int N;
	static Character[][] image;
	static String result = "";
	public static void find (int y, int x, int n ) {
		char color = image[y][x];
		for (int i = y; i < y + n; i++) {
			for (int j = x; j < x + n; j++) {
				if (color != image[i][j]) {
					int m = n/2;
					result += "(";
					find(y, x, m);
					find(y, x+m,m);
					find(y+m, x, m);
					find(y+m, x+m, m);
					result += ")";
					return;
				}
			}
		}
		if (color == '1') {
			result += "1";
		}
		else {
			result += "0";
		}
	}
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        image = new Character[N][N];
        for(int i = 0 ; i < N; i++) {
        	String S = br.readLine();
        	for (int j = 0 ; j < N; j++) {
        		image[i][j] = S.charAt(j);
        	}
        }
        find(0,0,N);
        System.out.println(result);
    }
}